app/models/manageiq/providers/amazon/cloud_manager/metrics_capture.rb

```diff
@@ -145,38 +154,47 @@ def counter_values_by_timestamp(metrics_by_counter_name)
         value   = i[:calculation].call(*metrics, interval)
 
         # For (temporary) symmetry with VIM API we create 20-second intervals.
+
         (last_ts + 20.seconds..ts).step_value(20.seconds).each do |inner_ts|
           counter_values_by_ts.store_path(inner_ts.iso8601, i[:vim_style_counter_key], value)
         end
       end
     end
+
     counter_values_by_ts
   end
 
   def metrics_by_counter_name(cloud_watch, counters, start_time, end_time)
     metrics_by_counter_name = {}
+
     counters.each do |c|
-      metrics = metrics_by_counter_name[c.metric_name] = {}
+      metrics = metrics_by_counter_name[c.metric_name] ||= {}
```

Comment:
So are some of these counters updating different counters?

Reply:
Metric can have a different set of dimensions, <br>depending on CloudWatch Agent settings:

![1 dimension sets](https://user-images.githubusercontent.com/7778414/55275563-51d03280-52e8-11e9-8c64-f45f4eb4ed82.png)


In my case, when they were changed:
*_(CloudWatch keeps 1 minute interval metrics for 2 weeks)_*

![2 memory 1](https://user-images.githubusercontent.com/7778414/55275568-5c8ac780-52e8-11e9-95e6-18dbc736691c.png)
![2 memory 2](https://user-images.githubusercontent.com/7778414/55275574-64e30280-52e8-11e9-9fb6-c943f39ed53c.png)
![2 memory 3](https://user-images.githubusercontent.com/7778414/55275575-6a404d00-52e8-11e9-8e60-fa9967e0ad73.png)


And dimensions can have different values too:

![3 pagefile 1](https://user-images.githubusercontent.com/7778414/55275579-788e6900-52e8-11e9-8996-9261496c98a4.png)
![3 pagefile 2](https://user-images.githubusercontent.com/7778414/55275584-7c21f000-52e8-11e9-8bfb-588dac9994f7.png)

Reply:
My thoughts:
- It makes no practical sense to collect same metric with different dimension sets at once (collect all dimensions in the same set instead, and aggregate it in required ways later). And I'm not sure, that's even possible. So, I'm not counting this case as worth handling.
- In case when settings were changed, we have two or more row series like:
```
config1 ->
  time1, dimension1, dimension2, value1
  time2, dimension1, dimension2, value2
config2 ->
  time3, dimension1,             value3
  time4, dimension1,             value4
  time5, dimension1,             value5
```
we just need to concatenate them together.

Reply:
- For Paging File(s) situation is bit more complicated: this counter provided for each file separately. Here we have two edge cases:
  - There is only one Paging File (which is most common scenario I believe), within 1 minute interval we have 2 or more values - we just take average of them.
  - There is more than one Paging File - to count common average we need to know each file's size, which is complicated within existing code design. Because this should be really rare case, I just left a note about that: https://github.com/ManageIQ/manageiq-providers-amazon/blob/9162f0482a9a5461d2784d61bdd0375cfbddc87b/app/models/manageiq/providers/amazon/cloud_manager/metrics_capture.rb#L33
I have thoughts about refactoring the whole metrics collection code, at least for AWS and Azure, which should make it more flexible and effective, and solve such issues too, but right now I want to be simple.